File: src/leetcode/java/bit_manipulation/SingleNumberII_137.java
package leetcode.java.bit_manipulation;

public class SingleNumberII_137 {
  public int singleNumber(int[] nums) {
    int one = 0x000, two = 0x000;
    for (int i = 0; i < nums.length; i++) {
      one = one ^ nums[i] & ~two;
      two = two ^ nums[i] & ~one;
    }
    return one;
  }
}

File: src/leetcode/python/bit_manipulation/SingleNumberII_137.py
class SingleNumberII_137(object):
    def singleNumber(self, nums):
        """
        :type nums: List[int]
        :rtype: int

        00, 01, 10 state machine
        """
        one = 0x0
        two = 0x0
        for num in nums:
            one = (one ^ num) & ~two
            two = (two ^ num) & ~one
        return one
